In [1]:
import pandas as pd
import random
from joblib import Parallel, delayed

## Import Datasets

#### Abstracts

In [3]:
abstract = pd.read_csv('./data/arxiv_data_210930-054931.csv')
abstract

FileNotFoundError: [Errno 2] No such file or directory: './data/arxiv_data_210930-054931.csv'

#### Articles

In [8]:
article1 = pd.read_csv('./data/articles1.csv')
article2 = pd.read_csv('./data/articles2.csv')
article3 = pd.read_csv('./data/articles3.csv')

In [16]:
article = pd.concat([article1,article2,article3],ignore_index=True)
article

Unnamed: 0      id                                              title  \
0                0   17283  House Republicans Fret About Winning Their Hea...   
1                1   17284  Rift Between Officers and Residents as Killing...   
2                2   17285  Tyrus Wong, ‘Bambi’ Artist Thwarted by Racial ...   
3                3   17286  Among Deaths in 2016, a Heavy Toll in Pop Musi...   
4                4   17287  Kim Jong-un Says North Korea Is Preparing to T...   
...            ...     ...                                                ...   
142565      146028  218078  An eavesdropping Uber driver saved his 16-year...   
142566      146029  218079  Plane carrying six people returning from a Cav...   
142567      146030  218080  After helping a fraction of homeowners expecte...   
142568      146031  218081  Yes, this is real: Michigan just banned bannin...   
142569      146032  218082  What happened in Washington state after voters...   

            publication                         author        date    year  \
0        New York Times                     Carl Hulse  2016-12-31  2016.0   
1        New York Times  Benjamin Mueller and Al Baker  2017-06-19  2017.0   
2        New York Times                   Margalit Fox  2017-01-06  2017.0   
3        New York Times               William McDonald  2017-04-10  2017.0   
4        New York Times                  Choe Sang-Hun  2017-01-02  2017.0   
...                 ...                            ...         ...     ...   
142565  Washington Post                       Avi Selk  2016-12-30  2016.0   
142566  Washington Post                  Sarah Larimer  2016-12-30  2016.0   
142567  Washington Post                    Renae Merle  2016-12-30  2016.0   
142568  Washington Post                 Chelsea Harvey  2016-12-30  2016.0   
142569  Washington Post           Christopher Ingraham  2016-12-29  2016.0   

        month                                                url  \
0        12.0                                                NaN   
1         6.0                                                NaN   
2         1.0                                                NaN   
3         4.0                                                NaN   
4         1.0                                                NaN   
...       ...                                                ...   
142565   12.0  https://web.archive.org/web/20161231004909/htt...   
142566   12.0  https://web.archive.org/web/20161231004909/htt...   
142567   12.0  https://web.archive.org/web/20161231004909/htt...   
142568   12.0  https://web.archive.org/web/20161231004909/htt...   
142569   12.0  https://web.archive.org/web/20161231004909/htt...   

                                                  content  
0       WASHINGTON  —   Congressional Republicans have...  
1       After the bullet shells get counted, the blood...  
2       When Walt Disney’s “Bambi” opened in 1942, cri...  
3       Death may be the great equalizer, but it isn’t...  
4       SEOUL, South Korea  —   North Korea’s leader, ...  
...                                                   ...  
142565           Uber driver Keith Avila picked up a p...  
142566           Crews on Friday continued to search L...  
142567       When the Obama administration announced a...  
142568       This story has been updated. A new law in...  
142569       The nation’s first recreational marijuana...  

[142570 rows x 10 columns]

#### Blogs

In [17]:
blog = pd.read_csv('./data/blogtext.csv')
blog

id gender  age              topic      sign          date  \
0       2059027   male   15            Student       Leo   14,May,2004   
1       2059027   male   15            Student       Leo   13,May,2004   
2       2059027   male   15            Student       Leo   12,May,2004   
3       2059027   male   15            Student       Leo   12,May,2004   
4       3581210   male   33  InvestmentBanking  Aquarius  11,June,2004   
...         ...    ...  ...                ...       ...           ...   
681279  1713845   male   23            Student    Taurus  01,July,2004   
681280  1713845   male   23            Student    Taurus  01,July,2004   
681281  1713845   male   23            Student    Taurus  01,July,2004   
681282  1713845   male   23            Student    Taurus  01,July,2004   
681283  1713845   male   23            Student    Taurus  01,July,2004   

                                                     text  
0                  Info has been found (+/- 100 pages,...  
1                  These are the team members:   Drewe...  
2                  In het kader van kernfusie op aarde...  
3                        testing!!!  testing!!!            
4                    Thanks to Yahoo!'s Toolbar I can ...  
...                                                   ...  
681279         Dear Susan,  I could write some really ...  
681280         Dear Susan,  'I have the second yeast i...  
681281         Dear Susan,  Your 'boyfriend' is fuckin...  
681282         Dear Susan:    Just to clarify, I am as...  
681283         Hey everybody...and Susan,  You might a...  

[681284 rows x 7 columns]

#### Movie

In [19]:
movie = pd.read_csv('./data/moviedataset.csv')
movie

Unnamed: 0        actors                               characters  \
0              0  Meryl Streep                         Clarissa Vaughan   
1              1  Meryl Streep                       Blue Mecha (voice)   
2              2  Meryl Streep                        Francesca Johnson   
3              3  Meryl Streep                         Miranda Priestly   
4              4  Meryl Streep                   Clara del Valle Trueba   
...          ...           ...                                      ...   
3918        3918  Will Ferrell                                      NaN   
3919        3919  Will Ferrell                                      NaN   
3920        3920  Will Ferrell                                    Angel   
3921        3921  Will Ferrell  Michael, Brian's Boyfriend (uncredited)   
3922        3922  Will Ferrell                                      NaN   

                        movie title                             genres  \
0                         The Hours                              Drama   
1      A.I. Artificial Intelligence  Adventure, Drama, Science Fiction   
2     The Bridges of Madison County                     Drama, Romance   
3             The Devil Wears Prada             Comedy, Drama, Romance   
4          The House of the Spirits                     Drama, Romance   
...                             ...                                ...   
3918             The Thin Pink Line                             Comedy   
3919              Men Seeking Women                    Comedy, Romance   
3920             Good Cop, Baby Cop                             Comedy   
3921                      Boat Trip                             Comedy   
3922    Bat Fight with Will Ferrell                                NaN   

      release year    IMDB ID  \
0             2002  tt0274558   
1             2001  tt0212720   
2             1995  tt0112579   
3             2006  tt0458352   
4             1994  tt0107151   
...            ...        ...   
3918          1998  tt0171825   
3919          1997  tt0219919   
3920          2007  tt1488565   
3921          2002  tt0285462   
3922          2009  tt1535001   

                                                Scripts  
0     Dearest, I feel certain that I am going mad ag...  
1     Al That was when the icecaps had melted becaus...  
2     - MICHAEL: Hi, sis. - CAROLYN: Hey. - CAROLYN:...  
3     Good luck. Hi. Uh, I have an appointment with ...  
4     THE HOUSE OF THE SPIRITS. It’s good to be back...  
...                                                 ...  
3918                                                NaN  
3919                                                NaN  
3920                                                NaN  
3921  Sol Kyung-Gu Moon So-Ri Kim Yeo-Jin Written an...  
3922                                                NaN  

[3923 rows x 8 columns]

#### Reddit

In [20]:
reddit = pd.read_csv('./data/Reddit_Data.csv')
reddit

clean_comment  category
0       family mormon have never tried explain them t...         1
1      buddhism has very much lot compatible with chr...         1
2      seriously don say thing first all they won get...        -1
3      what you have learned yours and only yours wha...         0
4      for your own benefit you may want read living ...         1
...                                                  ...       ...
37244                                              jesus         0
37245  kya bhai pure saal chutiya banaya modi aur jab...         1
37246              downvote karna tha par upvote hogaya          0
37247                                         haha nice          1
37248             facebook itself now working bjp’ cell          0

[37249 rows x 2 columns]

#### Songs

In [21]:
song = pd.read_csv('./data/song_lyrics.csv')
song

title      tag                    artist  year   views  \
0                Killa Cam      rap                   Cam'ron  2004  173166   
1               Can I Live      rap                     JAY-Z  1996  468624   
2        Forgive Me Father      rap                  Fabolous  2003    4743   
3             Down and Out      rap                   Cam'ron  2004  144404   
4                   Fly In      rap                 Lil Wayne  2005   78271   
...                    ...      ...                       ...   ...     ...   
5134851              Ocean      pop                   Effemar  2022       3   
5134852            64 Bars      rap                    Rapido  2022       4   
5134853    Raise Our Hands      pop  Culture Code, Pag & Mylo  2016       3   
5134854                CEO      rap               Antropolita  2022       5   
5134855         New Number  country         Alana Springsteen  2022       1   

                                             features  \
0                         {"Cam\\'ron","Opera Steve"}   
1                                                  {}   
2                                                  {}   
3        {"Cam\\'ron","Kanye West","Syleena Johnson"}   
4                                                  {}   
...                                               ...   
5134851                                            {}   
5134852                                            {}   
5134853            {Elex,"Culture Code / Pag & Mylo"}   
5134854                                            {}   
5134855                                            {}   

                                                    lyrics       id  \
0        [Chorus: Opera Steve & Cam'ron]\nKilla Cam, Ki...        1   
1        [Produced by Irv Gotti]\n\n[Intro]\nYeah, hah,...        3   
2        Maybe cause I'm eatin\nAnd these bastards fien...        4   
3        [Produced by Kanye West and Brian Miller]\n\n[...        5   
4        [Intro]\nSo they ask me\n"Young boy\nWhat you ...        6   
...                                                    ...      ...   
5134851  [Verse 1]\nDance for me now\nKeeping yourself ...  7882842   
5134852  [Intro]\n\nJa, ja\n\n[Part 1]\n\nR-A-H, Merhab...  7882843   
5134853  [Verse 1]\nHere our purpose feels alive\nWe ar...  7882845   
5134854  Jestem CEO w tym\nTo jara twoją bitch\nNikt na...  7882846   
5134855  [Verse 1]\nYou need a new number, one that ain...  7882848   

        language_cld3 language_ft language  
0                  en          en       en  
1                  en          en       en  
2                  en          en       en  
3                  en          en       en  
4                  en          en       en  
...               ...         ...      ...  
5134851            en          en       en  
5134852            de          de       de  
5134853            en          en       en  
5134854            pl          pl       pl  
5134855            en          en       en  

[5134856 rows x 11 columns]

#### Twitter

In [22]:
twitter = pd.read_csv('./data/Twitter_Data.csv')
twitter

clean_text  category
0       when modi promised “minimum government maximum...      -1.0
1       talk all the nonsense and continue all the dra...       0.0
2       what did just say vote for modi  welcome bjp t...       1.0
3       asking his supporters prefix chowkidar their n...       1.0
4       answer who among these the most powerful world...       1.0
...                                                   ...       ...
162975  why these 456 crores paid neerav modi not reco...      -1.0
162976  dear rss terrorist payal gawar what about modi...      -1.0
162977  did you cover her interaction forum where she ...       0.0
162978  there big project came into india modi dream p...       0.0
162979  have you ever listen about like gurukul where ...       1.0

[162980 rows x 2 columns]

## Create unified Dataset

In [27]:
dfs = {'abstract': abstract,
       'article': article,
       'blog': blog,
       'movie': movie,
       'reddit': reddit,
       'song': song,
       'twitter': twitter}

columns_to_keep = {'abstract': 'abstracts',
                   'article': 'content',
                   'blog': 'text',
                   'movie': 'Scripts',
                   'reddit': 'clean_comment',
                   'song': 'lyrics',
                   'twitter': 'clean_text'}  # Specify the column names to keep for each DataFrame

combined_data = pd.DataFrame(columns=['text', 'Source'])  # Initialize the combined DataFrame

for source, df in dfs.items():
    selected_column = df[columns_to_keep[source]]  # Select the specific column from the DataFrame
    combined_data = pd.concat([combined_data, pd.DataFrame({'text': selected_column, 'Source': source})], ignore_index=True)

In [28]:
combined_data

text    Source
0        Graph neural networks (GNNs) have been widely ...  abstract
1        Deep networks and decision forests (such as ra...  abstract
2        Graph convolutional networks (GCNs) are powerf...  abstract
3        With the increasing popularity of Graph Neural...  abstract
4        Machine learning solutions for pattern classif...  abstract
...                                                    ...       ...
6219038  why these 456 crores paid neerav modi not reco...   twitter
6219039  dear rss terrorist payal gawar what about modi...   twitter
6219040  did you cover her interaction forum where she ...   twitter
6219041  there big project came into india modi dream p...   twitter
6219042  have you ever listen about like gurukul where ...   twitter

[6219043 rows x 2 columns]

In [29]:
combined_data.to_csv('data.csv', index=False)

## Balancing the dataset

Targat, find the smallest group of sources and bring the others down to the same amount. 
We do not want the model to tend towards a category just because in the training phase more data from a specific category is being feed to the model


In [3]:
# loading the complex data generated above
data_origin = pd.read_csv('Data\data.csv')

In [4]:
# Calculate the minimum count among the sources
min_count = data_origin.groupby('Source').size().min()
min_count

3923

In [7]:
grouped_data = data_origin.groupby('Source')
grouped_data

In [8]:
# Initialize an empty list to store the balanced data
balanced_data = []
grouped_data = data_origin.groupby('Source')

for source, group in grouped_data:
    sampled_subset = group.sample(n=min_count, random_state=42)
    balanced_data.append(sampled_subset)

# Concatenate the balanced data from all sources
balanced_data = pd.concat(balanced_data)
# Reset the index of the balanced data
balanced_data = balanced_data.reset_index(drop=True)

# The 'balanced_data' DataFrame now contains exactly 3923 observations for each source variation
# with the 'text' and 'Source' features preserved


In [9]:
balanced_data

text    Source
0      Chest radiographs are among the most frequentl...  abstract
1      An explainable, efficient and lightweight meth...  abstract
2      Maritime domain is one of the most challenging...  abstract
3      Graph neural networks (GNNs) have attracted mu...  abstract
4      Automatic facial action unit (AU) recognition ...  abstract
...                                                  ...       ...
27456  brother modi dont conduct elections rather tha...   twitter
27457  travel india from has become such torturethank...   twitter
27458  pakistan and congressmen and pseudosecular one...   twitter
27459  wat our modijis photo has been printed thru ou...   twitter
27460  atleast this time vote shouldnt waste voted fo...   twitter

[27461 rows x 2 columns]

In [10]:
source_counts = balanced_data['Source'].value_counts()
print(source_counts)

abstract    3923
article     3923
blog        3923
movie       3923
reddit      3923
song        3923
twitter     3923
Name: Source, dtype: int64


In [11]:
balanced_data.to_csv('balenced_data.csv', index=False)

## Amount of Data
In this case, we have a dataset of 27,461 observations and we allocate 80% (approximately 21,968 observations) for training, it can be considered a reasonably large training set. However, it ultimately depends on the nature of your specific task and dataset. Discussion with the team might be needed

## Further steps needed
- Process the text data: Depending on the requirements and frame we set for the prject. Tokenization, lowercasing, stemming, lemmatization are needed. Can we use packages like NLTK?
- Vektorize the text Data: Converting the text data into numerical representations that a model can process. Techiniques are: BoW, TF-IDF or Word embeddings.
- Choose model: Naive Bayes, Support Vector Machines (SVM), Random Forest, or deep learning models like Convolutional Neural Networks (CNN) or Recurrent Neural Networks (RNN).
- Evaluation of the Model performance
- Possible fine tuning
- Testing